ARCs/arc-0023.md

```diff
@@ -0,0 +1,244 @@
+---
+arc: 23
+title: Append the hash of the contract description to the compiled application's bytes
+status: Draft
+---
+
+# Append the hash of the contract description to the compiled application bytes
+
+## Summary
+
+The following document introduces a convention for appending the hash of the contract description (as described in [ARC-4](https://github.com/algorandfoundation/ARCs/blob/main/ARCs/arc-0004.md)) to the compiled application's bytes.
+
+## Abstract
+
+The goal of this convention is to standardize the process of verifying and interacting with smart contracts.
+
+## Specification
+
+The key words "**MUST**", "**MUST NOT**", "**REQUIRED**", "**SHALL**", "**SHALL NOT**", "**SHOULD**", "**SHOULD NOT**", "**RECOMMENDED**", "**MAY**", and "**OPTIONAL**" in this document are to be interpreted as described in [RFC-2119](https://www.ietf.org/rfc/rfc2119.txt).
+
+> Comments like this are non-normative.
+
+## Overview
+
+By appending the hash of the contract description, any user with access to the blockchain could easily verify the ABI of an application and interact with it.
+
+> The goal is to have an off-chain ABI repository, where entries are indexed by the hash of the contract description.
+
+### Hash
+
+The hash is defined as follow:
+```
+SHA-512/256("arc0023" || <content of the metadata JSON file>)
+```
+In other words, it is the SHA-512/256 hash of the content of the metadata JSON file prefixed by the 7-byte string `arc0023`.
+### Appending
+
+The encoded object bytes are appended to the compiled application as a [bytecblock](https://developer.algorand.org/docs/get-details/dapps/avm/teal/opcodes/#bytecblock-bytes) containing 1 byte constant which is the encoded object.
```

Comment:
For programs that do not use `return`, I believe `pushbytes` is an issue since it will change the semantic.
For example the basic program:

```
#pragma version 5
int 1
```

changes semantic when adding the `pushbytes`

Reply:
Good point, perhaps the developer can choose between these options depending on whether they end with a `return` or not?

Reply:
I think the current assembler is going to get upset with a second bytecblock. I think that if it sees a bytecblock, it will not insert its own. I suppose that could be changed. Oh perhaps `pushbytes XXX; pop` (though that has the downside of costs 2 opcodes if you had planned on running off the end of the program with no opcodes to spare.

Reply:
@jannotti you're right, the assembler is ignoring its own `bytecblock`.
I guess the better solution is using `pushbytes` following a `return`. If `return` is already the last command then there's no need for another one of-course .

Reply:
It's weird, but this could also change semantics. It is currently ok to jump to the PC after the last byte as a way to end the program. I doubt anyone does it, but it's possible.